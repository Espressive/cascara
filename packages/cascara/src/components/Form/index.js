export { default, propTypes as FormPropTypes } from './Form';
