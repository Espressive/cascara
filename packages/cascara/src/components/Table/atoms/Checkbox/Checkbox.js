import React, { useCallback, useEffect, useRef } from 'react';
import pt from 'prop-types';
import styles from './Checkbox.module.scss';

const propTypes = {
  checked: pt.bool,
  indeterminate: pt.bool,
  name: pt.string,
  onChange: pt.func,
};

const Checkbox = ({
  checked = false,
  indeterminate = false,
  name,
  onChange = () => {},
}) => {
  const el = useRef();

  useEffect(() => {
    if (indeterminate) {
      // el.indeterminate = true;
    } else {
      // el.indeterminate = false;
    }
  }, [indeterminate]);

  const handleOnChange = useCallback(
    (e) => {
      const { checked, name } = e.currentTarget;

      onChange({ checked, name });
    },
    [onChange]
  );

  return (
    <input
      checked={checked ? 1 : 0}
      className={styles._}
      name={name}
      onChange={handleOnChange}
      ref={el}
      type={'checkbox'}
    />
  );
};

Checkbox.propTypes = propTypes;

export default Checkbox;
