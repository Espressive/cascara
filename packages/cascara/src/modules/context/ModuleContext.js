import { createContext } from 'react';

// We should ONLY put values here that are needed for ALL Modules
const defaultValue = {
  isEditing: false,
  isEditingAny: false,
  value: 'blue',
};

/*
 * We will use this baseline context to create any of our context providers from.
 * Any unique functionality should be included in that provider and not in this Context.
 */
const ModuleContext = createContext(defaultValue);

/*
 * We are also exporting the default value in case we need to hydrate or merge
 * any providers with these default values.
 */
export { defaultValue };

export default ModuleContext;
