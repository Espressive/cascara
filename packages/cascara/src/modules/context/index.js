export { default as ModuleContext } from './ModuleContext';
export { default as ModuleProvider } from './ModuleProvider';
