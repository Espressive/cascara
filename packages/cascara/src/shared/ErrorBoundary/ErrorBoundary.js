import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import DefaultErrorTemplate from './templates/DefaultErrorTemplate';

/**
 * A ErrorBoundary will catch and display an error
 */

class ErrorBoundary extends PureComponent {
  static propTypes = {
    children: PropTypes.element,
    errorTemplate: PropTypes.func,
  };

  static defaultProps = {
    children: null,
    errorTemplate: DefaultErrorTemplate,
  };

  state = {
    error: null,
    errorInfo: null,
  };

  // Catch errors in any components below and re-render with error message
  // You can also log error messages to an error reporting service here
  componentDidCatch = (error, errorInfo) => {
    this.setState({
      error: error,
      errorInfo: errorInfo,
    });
  };

  render() {
    const { children, errorTemplate } = this.props;

    const { error, errorInfo } = this.state;

    if (errorInfo) {
      // Error path
      return React.createElement(errorTemplate, {
        error,
        errorInfo,
      });
    }
    // Normally, just render children
    return children;
  }
}

export default ErrorBoundary;
