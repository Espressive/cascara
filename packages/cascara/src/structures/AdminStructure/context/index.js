export { default as AdminContext } from './AdminContext';
export { default as AdminProvider } from './AdminProvider';
