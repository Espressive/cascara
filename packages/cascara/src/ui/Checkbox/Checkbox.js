import React, { useEffect, useRef } from 'react';
import pt from 'prop-types';

const propTypes = {
  checked: pt.bool,
  indeterminate: pt.bool,
  name: pt.string,
  onChange: pt.func,
};

const Checkbox = ({
  checked = false,
  indeterminate = false,
  name,
  onChange = () => {},
}) => {
  const el = useRef();

  function handleOnChange(e) {
    const { checked, name } = e.currentTarget;

    onChange({ checked, name });
  }

  useEffect(() => {
    if (indeterminate) {
      // el.indeterminate = true;
    } else {
      // el.indeterminate = false;
    }
  }, [indeterminate]);

  return (
    <input
      checked={checked ? 1 : 0}
      name={name}
      onChange={handleOnChange}
      ref={el}
      type={'checkbox'}
    />
  );
};

Checkbox.propTypes = propTypes;

export default Checkbox;
