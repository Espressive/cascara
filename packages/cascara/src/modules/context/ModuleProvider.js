import React from 'react';
import pt from 'prop-types';

import ModuleContext, { defaultValue } from './ModuleContext';
import useToggle from '../../hooks/useToggle';

const propTypes = {
  children: pt.oneOfType([pt.element, pt.arrayOf(pt.element)]),
  value: pt.shape({
    isEditing: pt.bool,
  }),
};

const ModuleProvider = ({ children, value, ...props }) => {
  const [isEditing, setIsEditing] = useToggle(value?.isEditing || false);

  const mergedValues = {
    ...defaultValue,
    isEditing,
    setIsEditing,
    ...value,
  };

  return (
    <ModuleContext.Provider value={mergedValues} {...props}>
      {children}
    </ModuleContext.Provider>
  );
};

ModuleProvider.propTypes = propTypes;

export default ModuleProvider;
