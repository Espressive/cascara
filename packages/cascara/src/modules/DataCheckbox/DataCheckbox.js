import React, { useContext } from 'react';
import { Checkbox, useCheckboxState } from 'reakit/Checkbox';
import pt from 'prop-types';

import { ModuleContext } from '../context';
import styles from '../DataModule.module.scss';

import ErrorBoundary from '../../shared/ErrorBoundary';

const propTypes = {
  /** A module can have an Attribute, which will be used as form field name */
  attribute: pt.string,
  /** A Module can be defined to not present an editing state */
  isEditable: pt.bool,
  /** Presents the input without a label. NOT USER CONFIGURABLE */
  isLabeled: pt.bool,
  /** A Module needs to have a unique label relative to its context */
  label: pt.string,
  /** A Module can have a value */
  value: pt.bool,
};

const DataCheckbox = ({
  attribute,
  isEditable = true,
  isLabeled = true,
  label,
  value,
  ...rest
}) => {
  const { isEditing, formMethods } = useContext(ModuleContext);
  const checkbox = useCheckboxState({ state: Boolean(value) });

  const renderEditing = (
    <label htmlFor={label}>
      <Checkbox
        {...rest}
        {...checkbox}
        className={styles.Input}
        id={label}
        name={attribute || label}
        ref={formMethods?.register}
      />
      {label && isLabeled && <span className={styles.LabelText}>{label}</span>}
    </label>
  );

  const renderDisplay = (
    <span>
      <span
        className={styles.Input}
        data-checked={value ? true : undefined}
        {...rest}
      >
        {value}
      </span>
      {label && isLabeled && <span className={styles.LabelText}>{label}</span>}
    </span>
  );

  // Do not render an editable input if the module is not editable
  return (
    <ErrorBoundary>
      <div className={styles.Checkbox}>
        {isEditing && isEditable ? renderEditing : renderDisplay}
      </div>
    </ErrorBoundary>
  );
};

DataCheckbox.propTypes = propTypes;

export { propTypes };
export default DataCheckbox;
