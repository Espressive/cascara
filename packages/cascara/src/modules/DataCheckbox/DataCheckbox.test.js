import React from 'react';
import { fireEvent, render, screen, waitFor } from '@testing-library/react';

// import userEvent from '@testing-library/user-event';

import Form from '../../ui/Form';

import fixtures, {
  handleFormAction,
} from '../helpers/ModuleTestHelper.fixture';

const { checkboxTest } = fixtures;

describe('datacheckbox', () => {
  test('it returns the correct data type', async () => {
    const onAction = jest.fn();

    render(
      <Form
        actions={{
          modules: [
            {
              module: 'edit',
            },
          ],
        }}
        data={{
          id: '01',
          checkbox: false,
        }}
        dataDisplay={[
          {
            attribute: 'id',
            label: 'ID',
            module: 'text',
          },
          {
            attribute: 'checkbox',
            label: 'Checkbox',
            module: 'checkbox',
          },
        ]}
        isInitialEditing
        onAction={onAction}
      />
    );

    await waitFor(() => screen.findByRole('checkbox', { name: 'Checkbox' }));

    const checkboxInput = screen.getByRole('checkbox', { name: 'Checkbox' });
    fireEvent.click(checkboxInput);

    const saveButton = screen.getByRole('button', { name: 'Save' });
    fireEvent.click(saveButton);

    expect(onAction).toBeCalledWith({ name: 'edit.save' });
    // expect(log.mock.calls[0][1]).toEqual({ checkbox: true });
  });
});
