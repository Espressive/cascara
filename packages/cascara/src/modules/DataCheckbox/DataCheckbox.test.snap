// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataCheckbox display snapshot 1`] = `
<div>
  <div
    class="Checkbox"
  >
    <span>
      <span
        class="Input"
        data-checked="true"
      />
      <span
        class="LabelText"
      >
        Display
      </span>
    </span>
  </div>
</div>
`;

exports[`DataCheckbox editing snapshot 1`] = `
<div>
  <div
    class="Checkbox"
  >
    <label
      for="Editing"
    >
      <input
        aria-checked="false"
        class="Input"
        id="Editing"
        name="Editing"
        type="checkbox"
      />
      <span
        class="LabelText"
      >
        Editing
      </span>
    </label>
  </div>
</div>
`;
